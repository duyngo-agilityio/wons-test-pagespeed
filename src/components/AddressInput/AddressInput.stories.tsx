import type { Meta, StoryObj } from '@storybook/react';

// Components
import AddressInput from '.';

const meta: Meta<typeof AddressInput> = {
  title: 'Components/AddressInput',
  component: AddressInput,
  parameters: {
    layout: 'centered',
    controls: { expanded: true },
  },
};

export default meta;
type Story = StoryObj<typeof AddressInput>;

export const Default: Story = {
  args: {
    placeholder: 'Example placeholder',
    value: 'Example address',
  },
};
