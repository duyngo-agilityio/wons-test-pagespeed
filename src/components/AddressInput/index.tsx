'use client';

import { ChangeEvent, useEffect, useState } from 'react';
import { InputProps } from '@nextui-org/react';
import { useDebounce } from 'use-debounce';

// Icons
import { HiLocationMarker } from 'react-icons/hi';

// Components
import { Input, Text } from '@/components';
import { getLocationSuggestion } from '@/services';

interface AddressInputProps extends InputProps {}

interface LocationItem {
  properties: {
    formatted: string;
    name: string;
  };
}

const AddressInput = ({ placeholder = ' ', ...rest }: AddressInputProps) => {
  const [searchAddress, setSearchAddress] = useState('');
  const [locationsSuggestion, setLocationsSuggestion] = useState<
    LocationItem[]
  >([]);

  // Debounce searchAddress with a delay of 500ms
  const [debouncedSearchAddress] = useDebounce(searchAddress, 500);

  const handleOnChangeSearch = (event: ChangeEvent<HTMLInputElement>) => {
    const value = event.target.value;
    setSearchAddress(value);
  };

  useEffect(() => {
    const handleFetchData = async (keyword: string) => {
      const locationResponse = await getLocationSuggestion(keyword);

      setLocationsSuggestion(
        locationResponse?.features ? locationResponse.features : [],
      );
    };

    if (debouncedSearchAddress) {
      handleFetchData(debouncedSearchAddress);
    } else {
      setLocationsSuggestion([]);
    }
  }, [debouncedSearchAddress]);

  const handleClickSuggestion = (value: string) => {
    setSearchAddress(value);
    setLocationsSuggestion([]);
  };

  return (
    <div className="relative flex-1">
      <Input
        placeholder={placeholder}
        value={searchAddress}
        onChange={handleOnChangeSearch}
        endContent={<HiLocationMarker />}
        {...rest}
      />
      {locationsSuggestion.length !== 0 && (
        <div className="shadow-md p-3 absolute z-50 bg-white dark:bg-gray-400 w-full">
          {locationsSuggestion.map(({ properties }: LocationItem) => {
            const { formatted, name } = properties;

            return (
              <button
                key={name}
                className="flex w-full text-left py-[10px] px-[4px] gap-4 items-center relative cursor-pointer dark:text-white/80"
                onClick={() => handleClickSuggestion(formatted)}
              >
                <div className="w-[16px] h-[16px]">
                  <HiLocationMarker />
                </div>

                <Text text={formatted} className="dark:text-white/80" />
              </button>
            );
          })}
        </div>
      )}
    </div>
  );
};

export default AddressInput;
