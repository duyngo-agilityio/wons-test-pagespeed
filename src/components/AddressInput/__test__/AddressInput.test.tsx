import { fireEvent, render } from '@testing-library/react';

// Components
import AddressInput from '..';

jest.mock('next/navigation', () => ({
  ...jest.requireActual('next/navigation'),
  useSearchParams: jest.fn(() => ({
    get: () => () => 'John Deo',
  })),
}));

jest.mock('@/services', () => ({
  getLocationSuggestion: jest.fn().mockResolvedValue({
    features: [
      { properties: { formatted: '123 Main St', name: 'Location1' } },
      { properties: { formatted: '456 Elm St', name: 'Location2' } },
    ],
  }),
}));

describe('AddressInput component', () => {
  it('matches snapshot', () => {
    const { container } = render(
      <AddressInput onChange={jest.fn()} label="Address" />,
    );

    expect(container).toMatchSnapshot();
  });

  it('calls onChange when typing in input', () => {
    const handleChange = jest.fn();

    const { getByPlaceholderText } = render(
      <AddressInput
        label="Address"
        value=""
        onChange={handleChange}
        placeholder="Enter your address"
      />,
    );

    const inputElement = getByPlaceholderText('Enter your address');
    fireEvent.change(inputElement, { target: { value: 'New Address' } });

    expect(handleChange).toHaveBeenCalledTimes(1);
    expect(inputElement).toHaveValue('New Address');
  });
});
