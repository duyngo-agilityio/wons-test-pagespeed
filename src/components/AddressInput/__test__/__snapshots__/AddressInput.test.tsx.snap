// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddressInput component matches snapshot 1`] = `
<div>
  <div
    class="relative flex-1"
  >
    <div
      class="group flex flex-col data-[hidden=true]:hidden w-full relative justify-end data-[has-label=true]:mt-[29px]"
      data-filled="true"
      data-filled-within="true"
      data-has-elements="true"
      data-has-label="true"
      data-slot="base"
    >
      <div
        class="h-full flex flex-col"
        data-slot="main-wrapper"
      >
        <div
          class="relative w-full inline-flex tap-highlight-transparent flex-row items-center shadow-sm gap-3 min-h-10 transition-background motion-reduce:transition-none !duration-150 data-[hover=true]:bg-primary-100 text-primary group-data-[focus=true]:bg-primary-50 placeholder:text-primary outline-none group-data-[focus-visible=true]:z-10 group-data-[focus-visible=true]:ring-offset-background bg-gray-50 dark:bg-gray-600 hover:!bg-gray-200/50 dark:hover:!bg-gray-900 focus-within:!bg-gray-50 dark:focus-within:bg-gray-600 group-data-[focus=true]:!bg-gray-50 dark:group-data-[focus=true]:!bg-gray-600 h-12.5 p-3.75 rounded-[10px] group-data-[focus-visible=true]:ring-2 group-data-[focus-visible=true]:ring-offset-2 group-data-[focus-visible=true]:ring-blue-500"
          data-slot="input-wrapper"
          style="cursor: text;"
        >
          <label
            class="absolute pointer-events-none origin-top-left rtl:origin-top-right subpixel-antialiased block will-change-auto !duration-200 !ease-out motion-reduce:transition-none transition-[transform,color,left,opacity] text-primary group-data-[filled-within=true]:pointer-events-auto pb-0 z-20 -translate-y-1/2 group-data-[filled-within=true]:start-0 start-3 end-auto group-data-[filled-within=true]:-translate-y-[calc(100%_+_theme(fontSize.small)/2_+_20px)] pe-2 max-w-full text-ellipsis overflow-hidden group-data-[filled-within=true]:!text-blue-800 dark:group-data-[filled-within=true]:!text-white font-medium text-xl top-5"
            data-slot="label"
            for="react-aria-:r0:"
            id="react-aria-:r1:"
          >
            Address
          </label>
          <div
            class="inline-flex w-full items-center h-full box-border"
            data-slot="inner-wrapper"
          >
            <input
              aria-label="Address"
              aria-labelledby="react-aria-:r0: react-aria-:r1:"
              class="w-full font-normal !outline-none focus-visible:outline-none data-[has-start-content=true]:ps-1.5 data-[has-end-content=true]:pe-1.5 file:cursor-pointer file:bg-transparent file:border-0 autofill:bg-transparent bg-clip-text !text-blue-800/70 dark:!text-white/70 placeholder:text-blue-800/40 dark:placeholder:text-white/40 bg-transparent text-md"
              data-has-end-content="true"
              data-slot="input"
              id="react-aria-:r0:"
              placeholder=" "
              type="text"
              value=""
            />
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 20 20"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M5.05 4.05a7 7 0 119.9 9.9L10 18.9l-4.95-4.95a7 7 0 010-9.9zM10 11a2 2 0 100-4 2 2 0 000 4z"
                fill-rule="evenodd"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
